File: 2025/golang/day1/part2.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

type State struct {
	position  int
	zeroCount int
}

func (s *State) rotate(direction int, steps int) {
	// Avoid counting twice zero if starting at zero
	discount := 0
	if (s.position == 0) && (direction < 0) && (steps > 0) {
		discount = 1
	}

	newPosition := (s.position + direction*steps)
	zeroCount := floorDiv(newPosition, 100)
	newPosition = newPosition % 100
	// Adjust for when we crossed zero rotating left
	if newPosition < 0 {
		newPosition = 100 + newPosition
	}
	// Adjust sign of zeroCount
	if zeroCount < 0 {
		zeroCount *= -1
	}
	zeroCount -= discount
	// Adjst for when we cross zero rotating left and end up at zero
	if newPosition == 0 && direction < 0 {
		zeroCount++
	}

	s.position = newPosition
	s.zeroCount += zeroCount
}

func solve(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	defer file.Close()

	s := State{position: 50}
	direction := 1

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if first := line[0]; first == 'L' {
			direction = -1
		} else {
			direction = 1
		}

		steps, err := strconv.Atoi(line[1:])
		if err != nil {
			panic(err)
		}

		s.rotate(direction, steps)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Total:", s.zeroCount)
}

func floorDiv(a, b int) int {
	q := a / b
	r := a % b
	if r != 0 && (a^b) < 0 {
		q--
	}
	return q
}

func main() {
	// input := "input_test.txt" // Expected: 6
	input := "input.txt" // 6386
	solve(input)
}

File: 2025/cpp/day1/part2.cpp
#include <fstream>
#include <iostream>
#include <string>

using namespace std;


typedef struct {
    int position;
    int zero_count;
} State;


int floor_div(int a, int b) {
    int q = a / b;
    int r = a % b;
    if (r != 0 && (a ^ b) < 0) {
        q--;
    }

    return q;
}


void rotate(State *state, int direction, int steps) {
    int discount = 0;
    if (state->position == 0 && direction < 0 && steps > 0) {
        discount = 1;
    }
    int new_position = state->position + direction * steps;
    int zero_count = floor_div(new_position, 100);
    new_position = new_position % 100;

    if (new_position < 0) {
      new_position = 100 + new_position;
    }

    if (zero_count < 0) {
      zero_count *= -1;
    }
    zero_count -= discount;

    if (new_position == 0 && direction < 0) {
      zero_count++;
    }

    state->position = new_position;
    state->zero_count += zero_count;
}


int get_steps(const string &line) {
    return stoi(line.substr(1));
}


int main() {
    // string input_name = "input_test.txt"; // Expected output: 6
    string input_name = "input.txt"; // 6386
    ifstream file(input_name);

    int zero_count = 0;
    int position = 50;
    State state = {position, zero_count};
    int direction;
    int steps;
    string line;
    if (file.is_open()) {
        while (getline(file, line)) {
            if (line[0] == 'L') {
                direction = -1;
            } else {
                direction = 1;
            }
            steps = get_steps(line);

            rotate(&state, direction, steps);
        }

        file.close();
    } else {
        cerr << "Unable to open file: " << input_name << endl;
    }

    cout << "Total: " << state.zero_count << endl;

    return 0;
}
